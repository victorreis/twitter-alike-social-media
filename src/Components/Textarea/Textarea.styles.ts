import styled from 'styled-components';

import { toPx } from '../../Utils/Transform/toPx.util';
import { TextareaStyleProps } from './Textarea.types';

export const StyledTextarea = styled.textarea<TextareaStyleProps>`
  box-sizing: border-box;
  padding: 0.5rem;
  resize: none;
  flex: 1;

  border-width: 1px;
  border-color: ${({ theme }) => theme.colors.main.effect.normal}55;
  border-style: solid;
  border-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};

  ${({ theme }) => theme.typographies.body1};
  background-color: ${({ theme }) => theme.colors.background.default.lightest};
`;
