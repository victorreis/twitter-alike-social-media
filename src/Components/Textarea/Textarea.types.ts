import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredTextareaProps {
  /**
   * Value to be controlled by the component.
   */
  value: string;

  /**
   * Callback function that is called when the input is changed.
   */
  onChange: (newValue: string) => void;
}

export interface DefaultTextareaProps {
  /**
   * Controls when the textarea is resizable or not.
   * @default false
   */
  resizable?: boolean;

  /**
   * Controls the number of rows.
   * @default 4
   */
  rows?: number;
}

export interface OptionalTextareaProps {
  /**
   * Controls the maximum number of caracters allowed.
   * @default 777
   */
  maxLength?: number;
}

export type TextareaProps = RequiredTextareaProps &
  DefaultTextareaProps &
  OptionalTextareaProps &
  TestProps &
  Omit<
    React.HTMLAttributes<HTMLTextAreaElement>,
    'children' | 'onChange' | 'rows' | 'maxLength'
  >;

export type TextareaStyleProps = Required<DefaultTextareaProps>;
