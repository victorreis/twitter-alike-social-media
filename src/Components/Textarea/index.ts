export * from './Textarea.types';
export * from './Textarea';
