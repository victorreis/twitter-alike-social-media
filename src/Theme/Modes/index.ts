export { darkTheme } from './Dark.theme';
