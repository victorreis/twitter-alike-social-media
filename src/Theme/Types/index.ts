export * from './Borders.types';
export * from './Colors.types';
export * from './CustomTheme.types';
export * from './Sizes.types';
export * from './Typographies.types';
