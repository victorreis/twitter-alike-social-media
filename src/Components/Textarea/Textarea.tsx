import {
  forwardRef,
  ForwardRefRenderFunction,
  useState,
  useCallback,
  useEffect,
} from 'react';

import { TestProps } from '../../Config/Tests/Test.types';
import { StyledTextarea } from './Textarea.styles';
import { TextareaProps, DefaultTextareaProps } from './Textarea.types';

export const textareaDefaults: Required<DefaultTextareaProps> &
  Required<TestProps> = {
  testID: 'Textarea',
  resizable: false,
  rows: 4,
};

const TextareaComponent: ForwardRefRenderFunction<
  HTMLTextAreaElement,
  TextareaProps
> = (props, ref) => {
  const {
    testID = textareaDefaults.testID,
    value: initialValue,
    resizable = textareaDefaults.resizable,
    rows = textareaDefaults.rows,
    onChange,
    maxLength,
    ...others
  } = props;

  const [value, setValue] = useState(initialValue);

  useEffect(() => {
    setValue(() => initialValue);
  }, [initialValue]);

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLTextAreaElement>) => {
      const newValue = event.target.value;
      onChange(newValue);
      setValue(() => initialValue);
    },
    [initialValue, onChange]
  );

  return (
    <StyledTextarea
      ref={ref}
      data-testid={testID}
      maxLength={maxLength}
      onChange={handleChange}
      resizable={resizable}
      rows={rows}
      value={value}
      {...others}
    />
  );
};
export const Textarea = forwardRef(TextareaComponent);
