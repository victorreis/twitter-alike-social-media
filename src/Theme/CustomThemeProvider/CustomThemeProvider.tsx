/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
