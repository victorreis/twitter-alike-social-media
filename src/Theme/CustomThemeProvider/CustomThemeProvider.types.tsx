import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
