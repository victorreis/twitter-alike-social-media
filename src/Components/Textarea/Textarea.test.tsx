import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { toPx } from '../../Utils/Transform/toPx.util';
import { Textarea, textareaDefaults } from './Textarea';
import { RequiredTextareaProps, TextareaProps } from './Textarea.types';

describe('textarea component tests', () => {
  const value = 'text';
  const onChange = jest.fn();
  const newResizable = true;
  const newRows = 5;

  const requiredProps: RequiredTextareaProps = {
    value,
    onChange,
  };

  const setup = (props?: TextareaProps) => {
    const renderRTR = () =>
      renderRTRCreator<TextareaProps>(Textarea, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<TextareaProps>(Textarea, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(textareaDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(value);

      expect(element).toBeInTheDocument();
    });

    it(`should render '${String(
      textareaDefaults.resizable
    )}' as the default 'resizable'`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        resizable: textareaDefaults.resizable,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'resizable' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        resizable: newResizable,
      }).renderRTR().root;
      const element = instance.findByProps({ resizable: newResizable });

      expect(element).toBeTruthy();
    });

    it(`should render '${String(
      textareaDefaults.rows
    )}' as the default 'row'`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        rows: textareaDefaults.rows,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'row' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        rows: newRows,
      }).renderRTR().root;
      const element = instance.findByProps({ rows: newRows });

      expect(element).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(textareaDefaults.testID);

      expect(container).toHaveStyle({
        borderColor: `${themes.default.colors.main.effect.normal}55`,
        borderRadius: toPx(themes.default.borders.radius.LG),
        backgroundColor: themes.default.colors.background.default.lightest,
        ...themes.default.typographies.body1,
        // // color: hexToRgb(themes.default.colors.font.default),
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
