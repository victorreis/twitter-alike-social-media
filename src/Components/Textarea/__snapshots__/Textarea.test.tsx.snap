// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textarea component tests snapshot tests should render correctly 1`] = `
<textarea
  className="StyledTextarea-twitter-alike-social-media__sc-12f32i7-0 bivTbY"
  data-testid="Textarea"
  onChange={[Function]}
  rows={4}
  value="text"
/>
`;
